src/device/usbd.c

```diff
@@ -727,41 +703,59 @@ static bool process_set_config(uint8_t rhport, uint8_t cfg_num)
 
   while( p_desc < desc_end )
   {
-    // Each interface always starts with Interface or Association descriptor
+    tusb_desc_interface_assoc_t const * desc_itf_assoc = NULL;
+
+    // Class will always starts with Interface Association (if any) and then Interface descriptor
     if ( TUSB_DESC_INTERFACE_ASSOCIATION == tu_desc_type(p_desc) )
     {
-      p_desc = tu_desc_next(p_desc); // ignore Interface Association
-    }else
-    {
-      TU_ASSERT( TUSB_DESC_INTERFACE == tu_desc_type(p_desc) );
+      desc_itf_assoc = (tusb_desc_interface_assoc_t const *) p_desc;
+      p_desc = tu_desc_next(p_desc); // next to Interface
+    }
 
-      tusb_desc_interface_t const * desc_itf = (tusb_desc_interface_t const*) p_desc;
+    TU_ASSERT( TUSB_DESC_INTERFACE == tu_desc_type(p_desc) );
 
-      uint8_t drv_id;
-      uint16_t drv_len;
+    tusb_desc_interface_t const * desc_itf = (tusb_desc_interface_t const*) p_desc;
+    uint8_t drv_id;
+    uint16_t drv_len;
 
-      for (drv_id = 0; drv_id < USBD_CLASS_DRIVER_COUNT; drv_id++)
+    for (drv_id = 0; drv_id < USBD_CLASS_DRIVER_COUNT; drv_id++)
+    {
+      usbd_class_driver_t const *driver = &_usbd_driver[drv_id];
+
+      drv_len = 0;
+      if ( driver->open(rhport, desc_itf, &drv_len) )
       {
-        usbd_class_driver_t const *driver = &_usbd_driver[drv_id];
+        // Interface number must not be used already
+        TU_ASSERT( DRVID_INVALID == _usbd_dev.itf2drv[desc_itf->bInterfaceNumber] );
 
-        drv_len = 0;
-        if ( driver->open(rhport, desc_itf, &drv_len) )
+        TU_LOG2("  %s open\r\n", _usbd_driver[drv_id].name);
+        _usbd_dev.itf2drv[desc_itf->bInterfaceNumber] = drv_id;
+
+        // If IAD exist, assign all interfaces to the same driver
+        if (desc_itf_assoc)
         {
-          // Interface number must not be used already TODO alternate interface
-          TU_ASSERT( DRVID_INVALID == _usbd_dev.itf2drv[desc_itf->bInterfaceNumber] );
-          TU_LOG2("  %s open\r\n", _usbd_driver[drv_id].name);
-          _usbd_dev.itf2drv[desc_itf->bInterfaceNumber] = drv_id;
-          break;
+          // IAD's first interface number and class/subclass/protocol should match with opened interface
```

Comment:
The spec says:

```The values in these fields can be the same as the interface class, subclass and protocol values from any one of the associated interfaces```

not that it is the first one.

So, maybe change the comment to say that this is a TinyUSB requirement, not a spec requirement.

Reply:
You are right on the specs, but it does also say its prefer as well :) . I will keep this as it is, until there is a actual class that does differently. There is TU_ASSSERT(), we can always catch this error.
![image](https://user-images.githubusercontent.com/249515/79369103-a0478200-7f7a-11ea-9327-8c228effd39f.png)